### PR TITLE
Backport 9691 to release 3 6

### PR DESCRIPTION
**Pull request by mbernasocchi**
## Description
Backport #9691 to release 3.6

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by m-kuhn** [comment]
@signedav does that look good to you?

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 1]
```diff
@@ -2070,7 +2070,7 @@ namespace QgsWms
       wmsMaxWidth = std::max( wmsMaxWidthProj, wmsMaxWidthEnv );
     }
 
-    int width = this->width();
+    int width = mWmsParameters.getWidthAsInt();
```
Comment (signedav) [inline src/server/services/wms/qgswmsrenderer.cpp]: That won't be needed anymore when this #9727 is merged.

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 2]
```diff
@@ -2091,7 +2091,7 @@ namespace QgsWms
       wmsMaxHeight = std::max( wmsMaxHeightProj, wmsMaxHeightEnv );
     }
 
-    int height = this->height();
+    int height = mWmsParameters.getHeightAsInt();
```
Comment (signedav) [inline src/server/services/wms/qgswmsrenderer.cpp]: and this neither...

**Comment by mbernasocchi** [comment]
@andreasneumann we need to wait for #9727 to be merged in 3.6

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 3]
```diff
@@ -2053,21 +2053,51 @@ namespace QgsWms
...
     {
       return false;
     }
 
-    int wmsMaxHeight = QgsServerProjectUtils::wmsMaxHeight( *mProject );
+    //HEIGHT
+    int wmsMaxHeightProj = QgsServerProjectUtils::wmsMaxHeight( *mProject );
+    int wmsMaxHeightEnv = mSettings.wmsMaxHeight();
+    int wmsMaxHeight;
+    if ( wmsMaxWidthEnv != -1 && wmsMaxWidthProj != -1 )
```
Comment (m-kuhn) [inline src/server/services/wms/qgswmsrenderer.cpp]: @mbernasocchi looks like a leftover here. Can you fix that on master too?

```suggestion
    if ( wmsMaxHeightEnv != -1 && wmsMaxHeightProj != -1 )
```

File: tests/src/python/test_qgsserver_wms_getmap_size_project.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    # Set to True to re-generate reference files for this class
+    regenerate_reference = False
+
+    def setUp(self):
+        os.environ['QGIS_SERVER_WMS_MAX_WIDTH'] = '6000'
+        os.environ['QGIS_SERVER_WMS_MAX_HEIGHT'] = '6000'
+        super(TestQgsServerWMSGetMapSizeProject, self).setUp()
+        self.project = os.path.join(self.testdata_path, "test_project_with_size.qgs")
+        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="InvalidParameterValue">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
```
Comment (m-kuhn) [inline tests/src/python/test_qgsserver_wms_getmap_size_project.py]:
```suggestion
        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="Size error">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
```

File: tests/src/python/test_qgsserver_wms_getmap_size_server.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    # Set to True to re-generate reference files for this class
+    regenerate_reference = False
+
+    def setUp(self):
+        os.environ['QGIS_SERVER_WMS_MAX_WIDTH'] = '3000'
+        os.environ['QGIS_SERVER_WMS_MAX_HEIGHT'] = '3000'
+        super(TestQgsServerWMSGetMapSizeServer, self).setUp()
+        self.project = os.path.join(self.testdata_path, "test_project_with_size.qgs")
+        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="InvalidParameterValue">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
```
Comment (m-kuhn) [inline tests/src/python/test_qgsserver_wms_getmap_size_server.py]:
```suggestion
        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="Size Error">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
```

File: tests/src/python/test_qgsserver_wms_getmap_size_project.py (hunk 4 above)
Comment (mbernasocchi) [reply tests/src/python/test_qgsserver_wms_getmap_size_project.py]: Not sure about this one, the original test used InvalidParameterValue. @pblottiere?

File: tests/src/python/test_qgsserver_wms_getmap_size_project.py [hunk 6]
```diff
@@ -45,7 +45,7 @@ def setUp(self):
         os.environ['QGIS_SERVER_WMS_MAX_HEIGHT'] = '6000'
         super(TestQgsServerWMSGetMapSizeProject, self).setUp()
         self.project = os.path.join(self.testdata_path, "test_project_with_size.qgs")
-        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="InvalidParameterValue">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
+        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="Size error">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_getmap_size_project.py]: InvalidParameterValue is the correct value, "Size error" is not in the specs, IIRC this was fixed in master with https://github.com/qgis/QGIS/pull/9646

File: tests/src/python/test_qgsserver_wms_getmap_size_server.py [hunk 7]
```diff
@@ -43,7 +43,7 @@ def setUp(self):
         os.environ['QGIS_SERVER_WMS_MAX_HEIGHT'] = '3000'
         super(TestQgsServerWMSGetMapSizeServer, self).setUp()
         self.project = os.path.join(self.testdata_path, "test_project_with_size.qgs")
-        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="InvalidParameterValue">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
+        self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="Size error">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_getmap_size_server.py]: same as above, but I guess this mean that the source of the exception should be fixed too.

Comment (mbernasocchi) [reply tests/src/python/test_qgsserver_wms_getmap_size_server.py]: I tought so when I first saw it, but #9646 is not back ported to 3.6 if I see correctly?

Comment (elpaso) [reply tests/src/python/test_qgsserver_wms_getmap_size_server.py]: i.e. fix  `throw QgsBadRequestException( QStringLiteral( "Size error" )` in line 803 of src/server/services/wms/qgswmsrenderer.cpp

Comment (elpaso) [reply tests/src/python/test_qgsserver_wms_getmap_size_server.py]: yeah, no backports for #9646 , I think we must fix these minor issues one by one as they occour.

Comment (mbernasocchi) [reply tests/src/python/test_qgsserver_wms_getmap_size_server.py]: so should I fix the qgswmsrenderer.cpp and revert the change in the code?

**Review by elpaso — APPROVED** [review]
Looks good! Please just fix the error message code, so that we are not out of specs.

File: tests/src/python/test_qgsserver_wms_getmap_size_server.py (hunk 7 above)
Comment (elpaso) [reply tests/src/python/test_qgsserver_wms_getmap_size_server.py]: Yes, please, "Size error" is not an allowed value for WMS (SLD) specs, the correct one is `InvalidParameterValue`, I know it's not related to this PR but IMO is better to fix it than leave a bug and  wrong value in the test.

**Pull request merged**